Log when an existing question is added to a quiz.

JuanchoPestana opened the pull request:
Fixes #2664 

### Changes proposed in this Pull Request

* Log when an existing question is added to a quiz.
* Logs 'question_status' (value of status filter)
* Logs 'question_type' (value of type filter)
* Logs 'question_category_id' (ID of question category filter)
* Logs 'question_count' (number of questions being added to a quiz)

donnapep commented:
@JuanchoPestana I'm getting the following notices after adding a question:
```
PHP Notice:  Undefined index: question_status in /Users/donnapep/Local Sites/sensei/app/public/wp-content/plugins/sensei-lms/includes/class-sensei-lesson.php on line 2587

PHP Notice:  Undefined index: question_type in /Users/donnapep/Local Sites/sensei/app/public/wp-content/plugins/sensei-lms/includes/class-sensei-lesson.php on line 2588

PHP Notice:  Undefined index: question_category in /Users/donnapep/Local Sites/sensei/app/public/wp-content/plugins/sensei-lms/includes/class-sensei-lesson.php on line 2589
```

JuanchoPestana commented:
Hey @donnapep! I've tested a couple of times in different environments but I'm not getting those notices... The only way I got them is by commenting out the changes I did in the lesson-metadata.js file... Did you include the changes in this file when you tested? Please let me know if it's the case so that I can rule out that option... Thanks, and sorry for the inconvenience!

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2561,38 @@ public function lesson_add_existing_questions() {
 		}
 
+		// Log event: when an existing question is added to a quiz.
+		if ( $at_least_one_question_added ) {
```
Seems as though we can just check that `count( $questions ) > 0`.

JuanchoPestana replied:
Please see my comment below! (In the main thread...)

donnapep replied:
I still don't think we need both these variables though. Seems that we could do `count( $number_of_added_questions ) > 0` just as easily.

JuanchoPestana replied:
Done!

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2561,38 @@ public function lesson_add_existing_questions() {
 					if ( ! in_array( $quiz_id, $quizzes ) ) {
+
+						$at_least_one_question_added = true;
+						$number_of_added_questions++;
```
I don't think we need either of these new variables. See other comments.

JuanchoPestana replied:
Please see my comment below! (In the main thread...)

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2561,38 @@ public function lesson_add_existing_questions() {
+				'question_status'      => $question_data['question_status'],
+				'question_type'        => $question_data['question_type'] ? $question_data['question_type'] : 'all',
+				'question_category_id' => $question_data['question_category'] ? get_term_by( 'slug', $question_data['question_category'], 'question-category' )->term_id : 'all',
+				'question_count'       => $number_of_added_questions,
```
We can use `count( $questions )` as the questions count.

JuanchoPestana replied:
Please see my comment below! (In the main thread...)

donnapep commented:
> I've tested a couple of times in different environments but I'm not getting those notices...

Yup, looks like there must have been something amiss in my environment. I didn't have any issue testing it this time.

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2561,38 @@ public function lesson_add_existing_questions() {
-
-					add_post_meta( $question_id, '_quiz_question_order' . $quiz_id, $quiz_id . '000' . $question_count );
-					$question_type = Sensei()->question->get_question_type( $question_id );
+						add_post_meta( $question_id, '_quiz_question_order' . $quiz_id, $quiz_id . '000' . $question_count );
```
Is there a reason this was moved to inside the `if` statement? It seems a bit risky.

JuanchoPestana replied:
Please see my comment below! (In the main thread...)

JuanchoPestana commented:
Hi Donna,
Regarding the comments you left...

The `$questions` variable contains the array of questions received from the Ajax call. This means it contains ALL the questions the user selected with the checkbox and "tried" to add them to the quiz. 

The `if ( ! in_array( $quiz_id, $quizzes ) )` conditional makes sure that questions are not added twice. So thats why I created these variables:
- `$at_least_one_question_added`
- `$number_of_added_questions`

If we use `count($questions)` we would be logging the number of questions the user selected and tried to add, not the real number of questions added. (If this is the number you wish to log let me know and I'll gladly fix it, but I assumed we wanted to log the 'real' number of questions being added...)

If we use `count( $questions ) > 0` then we would be logging every time a user 'tries' to add a question, even if no questions were really added (because of the question being a duplicate). Again, if this is the desired behavior let me know; I assumed we wanted to log when questions are 'really' added.

This is also why I moved those lines inside the IF statement. Moving these lines fixed the following bug:
- Without the changes I implemented...
- Add a question to the quiz. 
- Add that same question again.
- The table that contains the quiz questions is updated with the duplicate question. (Now you have the same question twice)
- Refresh the page. The duplicate question disappears. 

It 'disappears' because the question was not added to the quiz. This line 

`add_post_meta( $question_id, '_quiz_id', $quiz_id, false );` 

is correctly placed inside the

 `if ( ! in_array( $quiz_id, $quizzes ) )` 

statement. With this change I made, now the question is NOT added to the table and also this 

`add_post_meta( $question_id, '_quiz_question_order' . $quiz_id, $quiz_id . '000' . $question_count );`

 is not added when it's not supposed to be added. This was only inserting that post meta value duplicately. 

Hope this answers your questions; if not, let me know and I'll try to explain myself better!

JuanchoPestana commented:
> A few comments. I'd also like to make a small tweak. Instead of calling the parameter question_category_id, I'd like to rename it to question_category.

Done!

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2561,38 @@ public function lesson_add_existing_questions() {
 					$quizzes = get_post_meta( $question_id, '_quiz_id', false );
+
 					if ( ! in_array( $quiz_id, $quizzes ) ) {
+
```
Can we get rid of this blank line on line 2567?  Also on line 2577.

JuanchoPestana replied:
Done!

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2560,36 @@ public function lesson_add_existing_questions() {
 					$quizzes = get_post_meta( $question_id, '_quiz_id', false );
+
 					if ( ! in_array( $quiz_id, $quizzes ) ) {
+						$at_least_one_question_added = true;
```
We don't need this anymore, right?

JuanchoPestana replied:
Missed it... Sorry for that!

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2547,6 +2547,9 @@ public function lesson_add_existing_questions() {
 
 		$return = '';
 
+		// Variables used for logging.
```
There's only one logging-related variable now.

donnapep commented on includes/class-sensei-lesson.php:
```diff
@@ -2557,23 +2560,35 @@ public function lesson_add_existing_questions() {
+			$event_properties = array(
+				'question_status'   => $question_data['question_status'],
+				'question_type'     => $question_data['question_type'] ? $question_data['question_type'] : 'all',
+				'question_category' => $question_data['question_category'] ? get_term_by( 'slug', $question_data['question_category'], 'question-category' )->term_id : 'all',
```
Just noticed this, but I see that `get_term_by` returns `(WP_Term|array|false)`. To be safe, we should check that the return value is not falsy before trying to access `term_id`.

JuanchoPestana replied:
Done! Let me know if you'd prefer a different approach or different default values...

donnapep commented:
We're not going to move forward with logging these details.
